File: lc-dc-sep/lc-dc-sep10/solution.py
# Problem: 2807. Insert Greatest Common Divisors in Linked List

class Solution:
    def insertGreatestCommonDivisors(self, head: Optional[ListNode]) -> Optional[ListNode]:
        def gcd(a, b):
            while b > 0:
                a, b = b, a % b
            return a
        
        curr = head
        while curr.next:
            val1, val2 = curr.val, curr.next.val
            curr.next = ListNode(gcd(val1, val2), curr.next)
            curr = curr.next.next
        return head

File: lc-dc-sep/lc-dc-sep10/solution.cpp
// Problem: 2807. Insert Greatest Common Divisors in Linked List

class Solution {
public:
    int gcd(int a, int b){
        while(b > 0){
            int t = a;
            a = b;
            b = t % b;
        }
        return a;
    }
    ListNode* insertGreatestCommonDivisors(ListNode* head) {
        ListNode* curr = head;
        while(curr->next){
            int val1 = curr->val, val2 = curr->next->val;
            curr->next = new ListNode(gcd(val1, val2), curr->next);
            curr = curr->next->next;
        }
        return head;
    }
};